pkg/controllers/rollout/controller.go

```diff
@@ -160,6 +169,54 @@ func (r *Reconciler) fetchLatestResourceSnapshot(ctx context.Context, crpName st
+// We currently just wait for those cluster to be cleanup so that we can have a clean slate to start compute the rollout plan.
+// TODO (rzhang): group all bindings pointing to the same cluster together when we calculate the rollout plan so that we can avoid this.
+func waitForResourcesToCleanUp(allBindings []*fleetv1beta1.ClusterResourceBinding, crp *fleetv1beta1.ClusterResourcePlacement) (bool, error) {
+	crpObj := klog.KObj(crp)
+	deletingBinding := make(map[string]bool)
+	bindingMap := make(map[string]*fleetv1beta1.ClusterResourceBinding)
+	// separate deleting bindings from the rest of the bindings
+	for _, binding := range allBindings {
+		if !binding.DeletionTimestamp.IsZero() {
+			deletingBinding[binding.Spec.TargetCluster] = true
+			klog.V(2).InfoS("Found a binding that is being deleted", "clusterResourcePlacement", crpObj, "binding", klog.KObj(binding))
+		} else {
+			if _, exist := bindingMap[binding.Spec.TargetCluster]; !exist {
+				bindingMap[binding.Spec.TargetCluster] = binding
+			} else {
+				return false, controller.NewUnexpectedBehaviorError(fmt.Errorf("the same cluster `%s` has bindings `%s` and `%s` pointing to it",
+					binding.Spec.TargetCluster, bindingMap[binding.Spec.TargetCluster].Name, binding.Name))
+			}
+		}
+	}
+	// check if there are any cluster that has a binding that is both being deleted and scheduled
+	for cluster, binding := range bindingMap {
+		// check if there is a  deleting binding on the same cluster
+		if deletingBinding[cluster] {
+			klog.V(2).InfoS("Find a binding assigned to a cluster with another deleting binding", "clusterResourcePlacement", crpObj, "binding", binding)
+			if binding.Spec.State == fleetv1beta1.BindingStateBound {
+				// the rollout controller won't move a binding from unscheduled to bound if there is a deleting binding on the same cluster.
+				return false, controller.NewUnexpectedBehaviorError(fmt.Errorf(
+					"find a cluster `%s` that has a bound binding `%s` and a deleting binding point to it", binding.Spec.TargetCluster, binding.Name))
+			}
+			if binding.Spec.State == fleetv1beta1.BindingStateUnscheduled {
+				// this is a very rare case that the resource was in the middle of being removed from a member cluster after it is unselected.
+				// then the cluster get selected and unselected in two scheduling before the member agent is able to clean up all the resources.
+				if binding.GetAnnotations()[fleetv1beta1.PreviousBindingStateAnnotation] == string(fleetv1beta1.BindingStateBound) {
+					// its previous state can not bound as rollout won't roll a binding with a deleting binding pointing to the same cluster.
+					return false, controller.NewUnexpectedBehaviorError(fmt.Errorf(
+						"find a cluster `%s` that has a unscheduled binding `%+s` with previous state is `bound` and a deleting binding point to it", binding.Spec.TargetCluster, binding.Name))
+				}
+				return true, nil
```

Comment:
Hi Ryan! In this case, if the rollout controller is set to wait, the unscheduled binding will never be deleted, am I understanding it correctly? The scheduler will not delete bindings

Reply:
yes, the unscheduled binding will only be deleted after the deleting binding disappears. However, there is an optimization that we can apply now that we know the previous state of an unscheduled binding to delete an unscheduled binding directly if it was never bound. I will leave this for next PR.

Reply:
How can the flow stuck though? The deleting binding will disappear on its own.

Reply:
Oh, right. I missed a part. It won't get stuck.